src/redux/observer/index.ts

```diff
@@ -18,105 +22,211 @@ import {
-  dispatch: Function
-  getState: Function
-}
-const subscribe = async ({ dispatch, getState }: ContractObserverParams) => {
-  const state = getState()
-
-  try {
-    contracts.registry = await getRegistryContract(state)
-    contracts.token    = await getTokenContract(state)
-    contracts.voting   = await getVotingContract(state)
-  } catch (err) {
-    dispatch(observerError(err))
-    return
+class ContractObserver {
+  private contracts: Map  = {}
+  private emitters: Map   = {}
+  private requests: Map   = {}
+  private responders: Map = {}
+  private callbacks: Map  = {}
+
+  constructor() {
+    this.respondToContractEvent = this.respondToContractEvent.bind(this)
   }
 
-  subscribeToContractEvent({
-    contract: contracts.registry,
-    eventName: APPLICATION_EVENT,
-    responder: applicationEventResponder(dispatch, getState),
-  })
-
-  subscribeToContractEvent({
-    contract: contracts.registry,
-    eventName: APPLICATION_WHITELISTED_EVENT,
-    responder: applicationWhitelistedEventResponder(dispatch, getState),
-  })
-
-  subscribeToContractEvent({
-    contract:  contracts.registry,
-    eventName: CHALLENGE_EVENT,
-    responder: challengeEventResponder(dispatch, getState),
-  })
-
-  subscribeToContractEvent({
-    contract:  contracts.registry,
-    eventName: CHALLENGE_SUCCEEDED_EVENT,
-    responder: challengeSucceededEventResponder(dispatch, getState),
-  })
-
-  subscribeToContractEvent({
-    contract:  contracts.registry,
-    eventName: CHALLENGE_FAILED_EVENT,
-    responder: challengeFailedEventResponder(dispatch, getState),
-  })
-
-  subscribeToContractEvent({
-    contract:  contracts.token,
-    eventName: APPROVAL_EVENT,
-    responder: approvalEventResponder(dispatch, getState),
-  })
-
-  subscribeToContractEvent({
-    contract:  contracts.token,
-    eventName: TRANSFER_EVENT,
-    responder: transferEventResponder(dispatch, getState),
-  })
-
-  subscribeToContractEvent({
-    contract:  contracts.voting,
-    eventName: VOTING_RIGHTS_GRANTED_EVENT,
-    responder: votingRightsGrantedEventResponder(dispatch, getState),
-  })
-
-  subscribeToContractEvent({
-    contract:  contracts.voting,
-    eventName: VOTE_COMMITTED_EVENT,
-    responder: voteCommittedEventResponder(dispatch, getState),
-  })
-}
+  private async cacheContracts(dispatch: Function, getState: Function): Promise<void> {
+    const state = getState()
+
+    try {
+      // TODO(geoff) refactor for concurrency
+      this.contracts.registry = await getRegistryContract(state)
+      this.contracts.token    = await getTokenContract(state)
+      this.contracts.voting   = await getVotingContract(state)
+    } catch (err) {
+      dispatch(observerError(err))
+    }
+  }
 
-const subscribeToContractEvent = ({ contract, eventName, responder }: Map): void => {
-  const emitter = contract.getEventEmitter(eventName) as EventEmitter
-  emitters[eventName] = emitter
-  emitter.on('data', responder)
-}
+  private registerContractEvents(): void {
```

Comment:
smells like a `registerContractEvents` method could exitst that takes `[contract, EVENT]`